File: tests/Support/build-test-db.php
<?php

/**
 * Build the SQLite test database by applying all InvoicePlane SQL migrations.
 *
 * Converts MySQL DDL/DML to SQLite-compatible syntax on the fly:
 *   - Strips ENGINE, DEFAULT CHARSET, COLLATE, AUTO_INCREMENT table options
 *   - Converts `backtick` identifiers to "double-quoted"
 *   - Converts AUTO_INCREMENT column option to AUTOINCREMENT on INTEGER PRIMARY KEY
 *   - Drops unsupported CHANGE/MODIFY/CHANGE COLUMN (emitted as no-ops)
 *   - Handles INSERT ... ON DUPLICATE KEY UPDATE → INSERT OR REPLACE
 *
 * Usage:
 *   php tests/Support/build-test-db.php [/path/to/database.sqlite]
 */
$dbPath = $argv[1] ?? __DIR__ . '/../../storage/test.sqlite';

$dir = dirname($dbPath);
if ( ! is_dir($dir)) {
    mkdir($dir, 0755, true);
}

if (file_exists($dbPath)) {
    unlink($dbPath);
}

$pdo = new PDO('sqlite:' . $dbPath);
$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
$pdo->exec('PRAGMA journal_mode=WAL; PRAGMA foreign_keys=OFF;');

$sqlDir = __DIR__ . '/../../application/modules/setup/sql';
$files  = glob($sqlDir . '/*.sql');
sort($files);

foreach ($files as $file) {
    $sql = file_get_contents($file);

    // Skip pure-comment/empty files
    $stripped = preg_replace('/\/\*.*?\*\//s', '', $sql);
    $stripped = preg_replace('/#[^\n]*/', '', $stripped);
    $stripped = trim($stripped);
    if ($stripped === '') {
        continue;
    }

    $statements = convertMySqlToSqlite($sql);

    foreach ($statements as $stmt) {
        $stmt = trim($stmt);
        if ($stmt === '' || $stmt === ';') {
            continue;
        }

        try {
            $pdo->exec($stmt);
        } catch (PDOException $e) {
            // Tolerate "duplicate column" and similar upgrade-idempotency errors
            $msg = $e->getMessage();
            if (
                str_contains($msg, 'duplicate column')
                || str_contains($msg, 'already exists')
                || str_contains($msg, 'no such column')
            ) {
                // ignore — migration ran on already-updated schema
                continue;
            }

            fwrite(STDERR, "ERROR in {$file}:\n  stmt: " . mb_substr($stmt, 0, 120) . "\n  err:  {$msg}\n");
        }
    }
}

// Seed minimum required data
seedDefaults($pdo);

echo "Test database built: {$dbPath}\n";

// ---------------------------------------------------------------------------

function convertMySqlToSqlite(string $sql): array
{
    // Remove MySQL # comments
    $sql = preg_replace('/^#[^\n]*/m', '', $sql);

    // Replace backtick identifiers with double-quoted
    $sql = preg_replace('/`([^`]+)`/', '"$1"', $sql);

    // Strip COMMENT '...' clauses (MySQL column/table comments)
    $sql = preg_replace('/\s+COMMENT\s+\'(?:[^\'\\\\]|\\\\.)*\'/i', '', $sql);

    // Strip AFTER "column" or AFTER column clauses (column ordering, not supported by SQLite)
    $sql = preg_replace('/\s+AFTER\s+(?:"[^"]+"|`[^`]+`|\w+)/i', '', $sql);

    // Remove CREATE TABLE options — handle both same-line and multi-line ENGINE/CHARSET clauses.
    // Replace )\n  ENGINE=...\n  CHARSET=...;  →  );
    $sql = preg_replace(
        '/\)(\s*(?:ENGINE\s*=\s*\S+|DEFAULT\s+CHARSET\s*=\s*\S+|COLLATE\s*=\s*\S+|AUTO_INCREMENT\s*=\s*\d+)\s*)+;/im',
        ');',
        $sql
    );
    // Fallback: handle any remaining options on the same ) line
    $sql = preg_replace(
        '/\)\s*(ENGINE\s*=\s*\S+)?\s*(DEFAULT\s+CHARSET\s*=\s*\S+\w*)?\s*(COLLATE\s*=\s*\S+)?\s*(AUTO_INCREMENT\s*=\s*\d+)?\s*;/i',
        ');',
        $sql
    );

    // Convert AUTO_INCREMENT column definition to INTEGER (SQLite INTEGER PRIMARY KEY auto-increments)
    $sql = preg_replace('/\bINT\b\s+(?:NOT NULL\s+)?AUTO_INCREMENT/i', 'INTEGER', $sql);
    $sql = preg_replace('/AUTO_INCREMENT/i', '', $sql);

    // Remove redundant KEY/INDEX definitions inside CREATE TABLE
    $sql = preg_replace('/,\s*(?:UNIQUE\s+)?KEY\s+"[^"]*"\s+\([^)]+\)/i', '', $sql);
    $sql = preg_replace('/,\s*INDEX\s+"[^"]*"\s+\([^)]+\)/i', '', $sql);
    // Also catch unquoted key names
    $sql = preg_replace('/,\s*(?:UNIQUE\s+)?KEY\s+\w+\s*\([^)]+\)/i', '', $sql);

    // Convert ALTER TABLE ... CHANGE [COLUMN] → no-op
    $sql = preg_replace(
        '/ALTER TABLE\s+"[^"]+"\s+CHANGE\s+(?:COLUMN\s+)?"[^"]+"\s+"[^"]+"\s+[^;]+;/i',
        '-- CHANGE COLUMN skipped;',
        $sql
    );

    // Convert ALTER TABLE ... MODIFY [COLUMN] → no-op
    $sql = preg_replace(
        '/ALTER TABLE\s+\S+\s+MODIFY\s+(?:COLUMN\s+)?"[^"]+"\s+[^;]+;/i',
        '-- MODIFY COLUMN skipped;',
        $sql
    );

    // Convert ENUM(...) → TEXT
    $sql = preg_replace('/\bENUM\s*\([^)]+\)/i', 'TEXT', $sql);

    // Convert LONGTEXT/MEDIUMTEXT/TINYTEXT → TEXT
    $sql = preg_replace('/\b(?:LONG|MEDIUM|TINY)TEXT\b/i', 'TEXT', $sql);

    // Convert VARCHAR(n) → TEXT
    $sql = preg_replace('/\bVARCHAR\s*\(\s*\d+\s*\)/i', 'TEXT', $sql);

    // Convert INT(n)/TINYINT(n)/SMALLINT(n)/BIGINT(n) → INTEGER
    $sql = preg_replace('/\b(?:TINY|SMALL|MEDIUM|BIG)?INT\s*\(\s*\d+\s*\)/i', 'INTEGER', $sql);
    $sql = preg_replace('/\bSMALLINT\b/i', 'INTEGER', $sql);
    $sql = preg_replace('/\bTINYINT\b/i', 'INTEGER', $sql);
    $sql = preg_replace('/\bBIGINT\b/i', 'INTEGER', $sql);

    // Convert DECIMAL(m,n) → REAL
    $sql = preg_replace('/\bDECIMAL\s*\(\s*\d+\s*,\s*\d+\s*\)/i', 'REAL', $sql);

    // INSERT ... ON DUPLICATE KEY UPDATE ... → INSERT OR REPLACE
    $sql = preg_replace(
        '/INSERT INTO\s+("([^"]+)")\s+\(([^)]+)\)\s+VALUES\s+\(([^)]+)\)\s+ON DUPLICATE KEY UPDATE[^;]+;/is',
        'INSERT OR REPLACE INTO $1 ($3) VALUES ($4);',
        $sql
    );

    // ---------------------------------------------------------------------------
    // Expand multi-clause ALTER TABLE into individual ALTER TABLE statements.
    // SQLite only supports one ADD COLUMN per statement, and does not support
    // ADD INDEX / ADD KEY at all (those become CREATE INDEX).
    // ---------------------------------------------------------------------------
    $sql = expandAlterTable($sql);

    // Split on semicolon + newline boundary
    $parts = preg_split('/;\s*\n/m', $sql);

    return array_map(fn (string $s) => trim($s), $parts);
}

/**
 * Split a multi-clause ALTER TABLE into individual SQLite-compatible statements.
 *
 * Input (MySQL):
 *   ALTER TABLE "t"
 *     ADD COLUMN "a" TEXT,
 *     ADD COLUMN "b" INTEGER,
 *     ADD INDEX "idx" ("a");
 *
 * Output (SQLite):
 *   ALTER TABLE "t" ADD COLUMN "a" TEXT;
 *   ALTER TABLE "t" ADD COLUMN "b" INTEGER;
 *   CREATE INDEX IF NOT EXISTS "idx" ON "t" ("a");
 */
function expandAlterTable(string $sql): string
{
    // Match: ALTER TABLE "name" <clauses>;
    return preg_replace_callback(
        '/ALTER\s+TABLE\s+("([^"]+)"|(\w+))\s+((?:[^;]|\n)+?)\s*;/is',
        static function (array $m): string {
            $table   = $m[1]; // quoted or bare table name
            $body    = $m[4];
            $clauses = preg_split('/,\s*(?=ADD\s)/i', $body);

            $out = [];
            foreach ($clauses as $clause) {
                $clause = trim($clause);
                if (preg_match('/^ADD\s+(?:PRIMARY\s+KEY|CONSTRAINT|CHECK)\b/i', $clause)) {
                    $out[] = '-- skipped: ' . preg_replace('/\s+/', ' ', $clause);
                } elseif (preg_match('/^ADD\s+(?:COLUMN\s+)?("[\w ]+"|\w+)\s+(INTEGER|TEXT|REAL|BLOB|NUMERIC|DATETIME|DATE|TIME|BOOLEAN|\w+\b.*)/is', $clause, $cm)) {
                    $out[] = "ALTER TABLE {$table} ADD COLUMN {$cm[1]} {$cm[2]}";
                } elseif (preg_match('/^ADD\s+(?:UNIQUE\s+)?(?:INDEX|KEY)\s+(?:"([^"]+)"|\w+)\s*(\([^)]+\))/is', $clause, $im)) {
                    $idxName = $im[1] ?? 'idx_' . md5($clause);
                    $cols    = $im[2];
                    $unique  = mb_stripos($clause, 'UNIQUE') !== false ? 'UNIQUE ' : '';
                    $out[]   = "CREATE {$unique}INDEX IF NOT EXISTS \"{$idxName}\" ON {$table} {$cols}";
                } else {
                    // Unknown clause — emit as comment
                    $out[] = '-- skipped: ' . preg_replace('/\s+/', ' ', $clause);
                }
            }

            return implode(";\n", $out) . ';';
        },
        $sql
    );
}

function seedDefaults(PDO $pdo): void
{
    // Check if ip_settings already has rows
    try {
        $count = $pdo->query('SELECT COUNT(*) FROM "ip_settings"')->fetchColumn();
        if ($count > 0) {
            return;
        }
    } catch (PDOException) {
        return;
    }

    $settings = [
        ['setting_key' => 'default_language',       'setting_value' => 'english'],
        ['setting_key' => 'currency_symbol',         'setting_value' => '$'],
        ['setting_key' => 'currency_symbol_placement', 'setting_value' => 'before'],
        ['setting_key' => 'date_format',             'setting_value' => 'Y-m-d'],
        ['setting_key' => 'time_format',             'setting_value' => 'H:i'],
        ['setting_key' => 'pdf_engine',              'setting_value' => 'pdfmake'],
        ['setting_key' => 'pdf_paper_size',          'setting_value' => 'a4'],
        ['setting_key' => 'pdf_paper_orientation',   'setting_value' => 'portrait'],
        ['setting_key' => 'read_only_toggle',        'setting_value' => 'paid'],
        ['setting_key' => 'next_invoice_number',     'setting_value' => '1'],
        ['setting_key' => 'next_quote_number',       'setting_value' => '1'],
        ['setting_key' => 'invoicenumber_prefix',    'setting_value' => ''],
        ['setting_key' => 'quotenumber_prefix',      'setting_value' => ''],
        ['setting_key' => 'disable_setup',           'setting_value' => '1'],
        ['setting_key' => 'sumex',                   'setting_value' => '0'],
    ];

    $stmt = $pdo->prepare('INSERT OR IGNORE INTO "ip_settings" ("setting_key","setting_value") VALUES (?,?)');
    foreach ($settings as $row) {
        $stmt->execute([$row['setting_key'], $row['setting_value']]);
    }

    // Seed a test user
    $pdo->exec("INSERT OR IGNORE INTO \"ip_users\"
        (\"user_name\",\"user_email\",\"user_password\",\"user_type\",\"user_active\")
        VALUES ('Admin','admin@test.local','" . password_hash('password', PASSWORD_DEFAULT) . "',1,1)");

    // Seed a default invoice group
    $pdo->exec("INSERT OR IGNORE INTO \"ip_invoice_groups\"
        (\"invoice_group_name\",\"invoice_group_identifier_format\",\"invoice_group_next_id\",\"invoice_group_left_pad\")
        VALUES ('Default','{{{id}}}',1,0)");
}

File: tests/Support/build-test-db.py
#!/usr/bin/env python3
"""
Build the SQLite test database from InvoicePlane MySQL migration files.

Converts MySQL DDL/DML to SQLite-compatible syntax.
"""

import re
import os
import sys
import glob
import sqlite3

DB_PATH = sys.argv[1] if len(sys.argv) > 1 else os.path.join(
    os.path.dirname(__file__), '..', '..', 'storage', 'test.sqlite'
)

SQL_DIR = os.path.join(os.path.dirname(__file__), '..', '..', 'application', 'modules', 'setup', 'sql')


def mysql_to_sqlite(sql: str) -> list[str]:
    """Convert a block of MySQL SQL to a list of SQLite-compatible statements."""

    # Remove /* ... */ block comments FIRST (before # removal, since # can appear inside them)
    sql = re.sub(r'/\*.*?\*/', '', sql, flags=re.DOTALL)
    # Remove MySQL-style comments (# ...)
    sql = re.sub(r'#[^\n]*', '', sql)

    # Replace backtick identifiers with plain names (SQLite uses them without quotes)
    sql = re.sub(r'`([^`]+)`', r'\1', sql)

    # -----------------------------------------------------------------------
    # CREATE TABLE transformations
    # -----------------------------------------------------------------------

    # Remove table-level KEY / INDEX lines inside CREATE TABLE (keep PRIMARY KEY)
    sql = re.sub(r',\s*(?:UNIQUE\s+)?(?:KEY|INDEX)\s+\w+\s*\([^)]+\)', '', sql, flags=re.IGNORECASE)

    # Convert AUTO_INCREMENT col type → INTEGER (SQLite INTEGER PK is autoincrement)
    sql = re.sub(r'INT\(\d+\)\s+NOT NULL\s+AUTO_INCREMENT', 'INTEGER NOT NULL', sql, flags=re.IGNORECASE)
    sql = re.sub(r'INT\(\d+\)\s+AUTO_INCREMENT', 'INTEGER', sql, flags=re.IGNORECASE)

    # Strip closing ENGINE/CHARSET/etc. options from CREATE TABLE
    sql = re.sub(
        r'\)\s*(?:ENGINE\s*=\s*\S+\s*)?(?:AUTO_INCREMENT\s*=\s*\d+\s*)?(?:DEFAULT\s+CHARSET\s*=\s*\S+\s*)?(?:COLLATE\s*=\s*\S+\s*)?;',
        ');',
        sql,
        flags=re.IGNORECASE
    )

    # -----------------------------------------------------------------------
    # Column type conversions
    # -----------------------------------------------------------------------
    sql = re.sub(r'\bLONGTEXT\b', 'TEXT', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bMEDIUMTEXT\b', 'TEXT', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bTINYTEXT\b', 'TEXT', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bVARCHAR\(\d+\)', 'TEXT', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bINT\(\d+\)', 'INTEGER', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bTINYINT\(\d+\)', 'INTEGER', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bSMALLINT\(\d+\)', 'INTEGER', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bBIGINT\(\d+\)', 'INTEGER', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bDECIMAL\(\d+,\s*\d+\)', 'REAL', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bFLOAT\(\d+,\s*\d+\)', 'REAL', sql, flags=re.IGNORECASE)
    sql = re.sub(r'\bDOUBLE\(\d+,\s*\d+\)', 'REAL', sql, flags=re.IGNORECASE)

    # Convert ENUM(...) → TEXT
    sql = re.sub(r"\bENUM\s*\([^)]+\)", 'TEXT', sql, flags=re.IGNORECASE)

    # -----------------------------------------------------------------------
    # ALTER TABLE transformations — split multi-ADD / multi-DROP into individual
    # -----------------------------------------------------------------------

    # Skip complex ALTER statements mixing ADD PRIMARY KEY with MODIFY (unsupported)
    sql = re.sub(
        r'ALTER TABLE\s+\w+\s+ADD\s+PRIMARY KEY\s*\([^)]+\)\s*,\s*MODIFY[^;]+;',
        '-- complex PK+MODIFY skipped;',
        sql,
        flags=re.IGNORECASE
    )

    # Remove ADD [UNIQUE] INDEX / ADD PRIMARY KEY clauses BEFORE multi-ADD split
    # so they don't interfere with the ADD COLUMN expansion
    sql = re.sub(r',\s*ADD\s+(?:UNIQUE\s+)?INDEX\s+\w+\s*\([^)]+\)', '', sql, flags=re.IGNORECASE)
    sql = re.sub(r',\s*ADD\s+PRIMARY KEY\s*\([^)]+\)', '', sql, flags=re.IGNORECASE)
    # Standalone ALTER TABLE ... ADD PRIMARY KEY ...
    sql = re.sub(r'ALTER TABLE\s+\w+\s+ADD\s+PRIMARY KEY\s*\([^)]+\)\s*;', '-- ADD PRIMARY KEY skipped;', sql, flags=re.IGNORECASE)

    sql = _expand_multi_add_column(sql)

    # Convert ALTER TABLE ... DROP col1, DROP col2, ... → individual DROPs
    sql = _expand_multi_drop_column(sql)

    # Remove AFTER <col> clause (SQLite doesn't support column ordering)
    sql = re.sub(r'\s+AFTER\s+\w+', '', sql, flags=re.IGNORECASE)

    # SQLite: ALTER TABLE ADD COLUMN col TEXT NOT NULL (without DEFAULT) is illegal.
    # Add a DEFAULT '' for TEXT NOT NULL and DEFAULT 0 for INTEGER NOT NULL.
    def add_default(m: re.Match) -> str:
        col_def = m.group(1)
        if re.search(r'\bDEFAULT\b', col_def, re.IGNORECASE):
            return m.group(0)  # already has DEFAULT
        if re.search(r'\bTEXT\b', col_def, re.IGNORECASE):
            return m.group(0).rstrip() + " DEFAULT ''"
        if re.search(r'\bINTEGER\b|\bREAL\b', col_def, re.IGNORECASE):
            return m.group(0).rstrip() + ' DEFAULT 0'
        return m.group(0) + " DEFAULT ''"

    sql = re.sub(
        r'(ALTER TABLE\s+\w+\s+ADD\s+COLUMN\s+\w+\s+\S+\s+NOT\s+NULL(?:\s+NOT\s+NULL)?)',
        add_default,
        sql,
        flags=re.IGNORECASE
    )

    # Convert ALTER TABLE ... CHANGE [COLUMN] old new ... → skip (unsupported)
    sql = re.sub(
        r'ALTER TABLE\s+\w+\s+CHANGE\s+(?:COLUMN\s+)?\w+\s+\w+\s+[^;]+;',
        '-- CHANGE COLUMN skipped;',
        sql,
        flags=re.IGNORECASE
    )

    # Convert ALTER TABLE ... MODIFY [COLUMN] ... → skip (unsupported)
    sql = re.sub(
        r'ALTER TABLE\s+\w+\s+MODIFY\s+(?:COLUMN\s+)?\w+\s+[^;]+;',
        '-- MODIFY COLUMN skipped;',
        sql,
        flags=re.IGNORECASE
    )

    # Convert INSERT ... ON DUPLICATE KEY UPDATE → INSERT OR REPLACE
    sql = re.sub(
        r'\bINSERT\s+INTO\b',
        'INSERT OR IGNORE INTO',
        sql,
        flags=re.IGNORECASE
    )
    sql = re.sub(
        r'\s+ON\s+DUPLICATE\s+KEY\s+UPDATE\s+[^;]+;',
        ';',
        sql,
        flags=re.IGNORECASE
    )

    # -----------------------------------------------------------------------
    # UPDATE: skip MySQL-specific function forms
    # -----------------------------------------------------------------------
    # Must run AFTER ALTER TABLE expansion so DOTALL doesn't consume preceding DDL
    sql = re.sub(r'UPDATE\s+\w+\s+SET\s+\w+\s*=\s*CONCAT\s*\([^;]+;', '-- CONCAT update skipped;', sql, flags=re.IGNORECASE | re.DOTALL)
    sql = re.sub(r'UPDATE\s+\w+\s+SET\s+\w+\s*=\s*CASE\s[^;]+;', '-- CASE update skipped;', sql, flags=re.IGNORECASE | re.DOTALL)

    # Split into statements
    parts = re.split(r';\s*\n', sql)
    return [p.strip() for p in parts if p.strip() and p.strip() != '--']


def _expand_multi_drop_column(sql: str) -> str:
    """
    SQLite only supports one DROP COLUMN per ALTER TABLE (SQLite 3.35+).
    Split: ALTER TABLE t DROP a, DROP b; → two statements.
    """
    def expand_drop(m: re.Match) -> str:
        table = m.group(1)
        drops = re.findall(r'DROP\s+(?:COLUMN\s+)?(\w+)', m.group(2), re.IGNORECASE)
        return ';\n'.join(f'ALTER TABLE {table} DROP COLUMN {col}' for col in drops)

    return re.sub(
        r'ALTER TABLE\s+(\w+)\s+(DROP\s+\w+(?:\s*,\s*DROP\s+\w+)+)',
        expand_drop,
        sql,
        flags=re.IGNORECASE
    )


def _expand_multi_add_column(sql: str) -> str:
    """
    SQLite only supports one ADD COLUMN per ALTER TABLE.
    Split: ALTER TABLE t ADD COL a ..., ADD COL b ...; → two statements.
    Also handles: ALTER TABLE t ADD a INT, ADD b INT; (without COLUMN keyword).
    """

    def expand_match(m: re.Match) -> str:
        table = m.group(1)
        adds_block = m.group(2)
        # Split on commas that precede ADD
        parts = re.split(r',\s*(?=ADD\b)', adds_block, flags=re.IGNORECASE)
        stmts = []
        for part in parts:
            part = part.strip()
            if re.match(r'^ADD\b', part, re.IGNORECASE):
                # Ensure ADD COLUMN keyword
                part = re.sub(r'^ADD\s+(?!COLUMN\b)', 'ADD COLUMN ', part, flags=re.IGNORECASE)
                stmts.append(f'ALTER TABLE {table} {part}')
        return ';\n'.join(stmts) if stmts else m.group(0)

    sql = re.sub(
        r'ALTER TABLE\s+(\w+)\s+(ADD(?:\s+COLUMN)?\s+\w[^;]+(?:,\s*ADD[^;]+)+)',
        expand_match,
        sql,
        flags=re.IGNORECASE
    )
    return sql


def build_db(db_path: str) -> None:
    os.makedirs(os.path.dirname(os.path.abspath(db_path)), exist_ok=True)
    if os.path.exists(db_path):
        os.unlink(db_path)

    con = sqlite3.connect(db_path)
    con.execute('PRAGMA journal_mode=WAL')
    con.execute('PRAGMA foreign_keys=OFF')

    files = sorted(glob.glob(os.path.join(SQL_DIR, '*.sql')))

    for filepath in files:
        with open(filepath) as f:
            raw = f.read()

        statements = mysql_to_sqlite(raw)

        for stmt in statements:
            if not stmt or stmt.startswith('--'):
                continue
            try:
                con.execute(stmt)
            except sqlite3.OperationalError as e:
                msg = str(e)
                # Tolerate idempotency errors from re-running migrations
                if any(k in msg for k in ('duplicate column', 'already exists', 'no such column', 'no such table')):
                    pass
                else:
                    print(f'ERROR in {os.path.basename(filepath)}:', file=sys.stderr)
                    print(f'  stmt: {stmt[:150]}', file=sys.stderr)
                    print(f'  err:  {msg}', file=sys.stderr)

    con.commit()
    seed_defaults(con)
    con.commit()
    con.close()
    print(f'Test database built: {db_path}')


def seed_defaults(con: sqlite3.Connection) -> None:
    count = con.execute('SELECT COUNT(*) FROM ip_settings').fetchone()[0]
    if count > 0:
        return

    settings = [
        ('default_language',        'english'),
        ('currency_symbol',         '$'),
        ('currency_symbol_placement','before'),
        ('date_format',             'Y-m-d'),
        ('time_format',             'H:i'),
        ('pdf_engine',              'pdfmake'),
        ('pdf_paper_size',          'a4'),
        ('pdf_paper_orientation',   'portrait'),
        ('read_only_toggle',        'paid'),
        ('next_invoice_number',     '1'),
        ('next_quote_number',       '1'),
        ('invoicenumber_prefix',    ''),
        ('quotenumber_prefix',      ''),
        ('disable_setup',           '1'),
        ('sumex',                   '0'),
    ]
    for key, val in settings:
        con.execute(
            'INSERT OR IGNORE INTO ip_settings (setting_key, setting_value) VALUES (?,?)',
            (key, val)
        )

    import hashlib, hmac
    # Simple password hash placeholder — tests won't actually authenticate through CI3 login
    con.execute(
        "INSERT OR IGNORE INTO ip_users (user_name,user_email,user_password,user_type,user_active) "
        "VALUES ('Admin','admin@test.local','$2y$10$testhashedpassword.notreal',1,1)"
    )

    con.execute(
        "INSERT OR IGNORE INTO ip_invoice_groups "
        "(invoice_group_name,invoice_group_identifier_format,invoice_group_next_id,invoice_group_left_pad) "
        "VALUES ('Default','{{{id}}}',1,0)"
    )


if __name__ == '__main__':
    build_db(DB_PATH)
